packages/frontend/src/components/connect-wallet/index.tsx

```diff
@@ -34,6 +46,85 @@ export const ConnectWallet = ({ mode }: ConnectWalletProps) => {
     const [connectPopoverOpen, setConnectPopoverOpen] = useState(false);
     const [accountPopoverOpen, setAccountPopoverOpen] = useState(false);
 
+    const [wrongChainModalOpen, setWrongChainModalOpen] = useState(false);
+    const [canFreelySwitchNetworks, setCanFreelySwitchNetworks] =
+        useState(false);
+
+    // this effect saves to the internal state the target chain as taken from the
+    // url search params.
+    // it doesn't do that if the user can freely switch networks (i.e. if the user
+    // has already landed on the page)
+    useEffect(() => {
+        if (canFreelySwitchNetworks) return;
+        const targetChainName = searchParams.get("chain");
+        if (!targetChainName) return;
+        const targetChain = chains.find(
+            (c) => c.name.toLowerCase() === targetChainName.toLowerCase()
+        );
+        if (!targetChain) return;
+        setChainFromSearchParams(targetChain);
+    }, [canFreelySwitchNetworks, chains, searchParams]);
+
+    // this effect does a couple things:
+    // - in case the dapp's url has no chain search param, it sets it to the currently connected chain
+    // - if a target chain is specified, but we're in the middle of a network switching event, it does nothing
+    // - if a target chain is specified, it is different to the currently active chain, and we're not in the
+    // middle of a network switching event, it shows the wrong network modal, inviting the user to switch the
+    // network in their wallet (it does so only the first time)
+    useEffect(() => {
```

Comment:
it would be great if we could break this big `useEffect`

Reply:
I think in the latest push I've done that, I've overall simplified the logic a bit